### ISSUE TITLE
Not all emoji sequences recommended for general interchange (RGI) cluster

### ISSUE THREAD

**Issue by rsheeter** [harfbuzz/harfbuzz#3017]
In testing emoji rgi clustering it seems polar bears and regional flags don't cluster. https://github.com/rsheeter/hb-emoji-clusters/blob/main/try_shape-stdout.txt has a list. For context, I tested all the sequences in https://unicode.org/Public/emoji/14.0/emoji-test.txt, which enumerates emoji sequences.

IIUC per Mark Davis all emoji should be grapheme clusters. I thought that would mean HB would cluster them but seemingly not. I see discussion on #2265 where a fix to make it so was discussed.

In the event emoji were to span multiple files it would help Chrome itemization if emoji rgi consistently formed clusters. If it's desirable that they don't I'd appreciate if someone could ELI5.

**Comment by behdad** [comment]
Regional-indicators are Complicated(TM) as seen in #2265.

Polar bear is the weirdest thing I've seen in Unicode:  🐻 U+1F43B, ZWJ U+200D,❄ U+2744.
️
This is our code implementing the grapheme logic: https://github.com/harfbuzz/harfbuzz/blob/bd5502f09f09b3f464754395880f9db30a5fcf60/src/hb-ot-shape.cc#L466-L517

For emoji, we append any ZWJ,Extended_Pictograph sequence to the previous cluster. U+2744 SNOWFLAKE *is* in that list. So I expect that we handle this sequence correctly. Let me check.

**Comment by behdad** [comment]
Maybe we should add an equivalent of your code to the test suite.

**Comment by behdad** [comment]
Oops... Bad bug in emoji table generator...

**Comment by behdad** [comment]
@rsheeter Can you please rerun your script against master and attach conclusion?

**Comment by rsheeter** [comment]
It would appear you have rescued the polar bear! https://github.com/rsheeter/hb-emoji-clusters/commit/85bfd1fe44434b1d69305082e335a48401a63506

**Comment by behdad** [comment]
Re the regional-indicator pairs: https://github.com/harfbuzz/harfbuzz/issues/2265#issuecomment-858191957

**Comment by rsheeter** [comment]
#3018 probably fixes this;.

**Comment by behdad** [comment]
> #3018 probably fixes this;.

Interesting to see if it actually does.

**Comment by rsheeter** [comment]
`0 / 4702 failures`, looks like it does :)

**Comment by drott** [comment]
Big thanks for doing this list, @rsheeter! And glad we found an actual issue here.

**Comment by khaledhosny** [comment]
This should be closed now, right?

**Comment by behdad** [comment]
Is fixed. Importing as a test would be nice.

**Comment by behdad** [comment]
Here's a slow way to get started. It's slow because we don't have `--unicode-files`...

```
$ echo x > null; curl https://www.unicode.org/Public/emoji/13.1/emoji-zwj-sequences.txt | cut -d';' -f1 | cut -d'#' -f1 | grep . | while read line; do ./hb-shape --font-file=null --unicodes="$line" --no-positions | grep -q '=[^0]' && ./hb-shape --font-file=null --unicodes="$line" --no-positions --show-text --show-unicode; done
```

Maybe just make `gen-emoji-table.py` emit a test file somewhere in `test/shaping/data/in-house/tests`. The need for non-empty null font file is also because of https://github.com/harfbuzz/harfbuzz/issues/2567
